errors.yaml

```
@@ -1,29 +1,33 @@
 ---
 errors:
   - code: index_creation_failed
-    description: "An error occurred while trying to create an index, check out our guide on [index creation](https://docs.meilisearch.com/learn/core_concepts/indexes.html)"
+    description: "An error occurred while trying to create an index, check out our guide on [index creation](https://docs.meilisearch.com/learn/core_concepts/indexes.html)."
   - code: index_already_exists
-    description: "An index with this UID already exists, check out our guide on [index creation](https://docs.meilisearch.com/learn/core_concepts/indexes.html)"
+    description: "An index with this UID already exists, check out our guide on [index creation](https://docs.meilisearch.com/learn/core_concepts/indexes.html)."
   - code: index_not_found
-    description: "An index with this UID was not found, check out our guide on [index creation](https://docs.meilisearch.com/learn/core_concepts/indexes.html)"
+    description: "An index with this UID was not found, check out our guide on [index creation](https://docs.meilisearch.com/learn/core_concepts/indexes.html)."
   - code: invalid_index_uid
-    description: "There is an error in the provided index format, check out our guide on [index creation](https://docs.meilisearch.com/learn/core_concepts/indexes.html)"
+    description: "There is an error in the provided index format, check out our guide on [index creation](https://docs.meilisearch.com/learn/core_concepts/indexes.html)."
   - code: index_not_accessible
     description: "An internal error occurred while trying to access the requested index."
   - code: invalid_state
-    description: "An error that occurs when the database is in an invalid state. Deleting the database and re-indexing should solve the problem."
+    description: "The database is in an invalid state. Deleting the database and re-indexing should solve the problem."
   - code: missing_primary_key
     description: "All documents must be identified with a unique key. MeiliSearch will try to infer the document field to use it as a primary key. If it can't infer it, the primary key must be specified manually, and this error is returned. See  our [guide on primary keys](https://docs.meilisearch.com/learn/core_concepts/documents.html#structure)."
+  - code: primary_key_inference_failed 
+    description: "The provided document does not contain any field with the substring `id`. Please add `id` to one of your fields so it can be used as the primary key during inference, or [manually designate the primary key](/learn/core_concepts/documents.md#setting-the-primary-key)."
```

Honest question: should the first `field` be plural or singular? Plural feels more correct to my ear, but I'm really not sure.

A second comment: isn't it better to recommend users to manually set the primary key field first? Seems like the easier solution since it doesn't require the dev to mess around with the dataset.

A third and very important change: as far as I understand, `errors.yaml` is not parsed in the same way as `*.md`, meaning that links need to be hardcoded URLs to the compiled vuepress output. Can you confirm that, @dichotommy?

```suggestion
    description: "The provided document does not contain any fields with the substring `id`. [Manually designate the primary key](https://docs.meilisearch.com/learn/core_concepts/documents.md#setting-the-primary-key) or add `id` to one of your fields so it can be used as the primary key during inference."
```

---

> isn't it better to recommend users to manually set the primary key field first? Seems like the easier solution since it doesn't require the dev to mess around with the dataset.

Do we need to suggest that in the error message? Or can we do that [here](https://docs.meilisearch.com/learn/core_concepts/documents.html#primary-field)?

---

1. I believe the word field here can be singular or plural. Could be wrong OFC. Probably better just to rephrase it:

```suggestion
    description: "The first provided document contains no fields with the substring `id`. [Manually designate the primary key](https://docs.meilisearch.com/learn/core_concepts/documents.md#setting-the-primary-key) or add `id` to one of your fields so it can be used as the primary key during inference."
```

2. I could be wrong here, but I took the addition of this error as part of a move towards primary key inference.

The current distinction between manual and inferred primary keys is difficult to explain to users. It's unclear what the benefits are of manually setting a primary key vs. allowing it to be inferred. There are two possible solutions to this that I see: either committing to inference, or committing to manual primary keys.

I believe we are heading in the direction of removing manual primary keys, and instead going full inference, i.e. you're fine if your dataset already contains a field with the substring `id`, and otherwise you should add one for MeiliSearch to use. This obviously would introduce some friction to using MeiliSearch (requires the extra step of adding an `id` field to some datasets) but would greatly clarify the documentation/understanding side, since users would know exactly what to expect and, when sharing datasets (e.g. for troubleshooting), the primary key would be much easier to identify.

@gmourier Which direction is the core product team leaning right now?

3. I believe you are correct Gui! I haven't tested it myself, but the fact that all links currently in `errors.yaml` are hardcoded suggests to me that it is the case.

---

Very good question @dichotommy, thanks for asking! 

I personally believe that our inference process is not good today for some specific cases. Analytics will allow us to verify the hypotheses that we have. 

For example, if the majority of people use `id` we could completely remove the inference and default to the `id` field when no primary key has been set on the index in the first place.

We could also strictly expect an `id` attribute as most databases do today and generate a unique value when the users do not provide it in their documents.

This adds friction in a way but this friction will be quickly reached and allows to solve other hidden frictions.

For example: 

The user has no `id` field in his document, we generate it automatically and he can have a first coherent and functional search whereas today if the inference is missed and we choose a field containing `id` before `id` (which it's the primary key that he wanted at first), he ends up with a completely inconsistent first search.

The Aha moment is gone and he won't see all the value we could have brought directly in his first use of the product.

If he hasn't gone elsewhere and insists, this is what happens: (The condition is that he must have understood by himself that the search is not in a good shape because of a bad inference)

He will have to check the document, find how to delete an index, delete it, recreate it, find that he can set a primary key to disable the inference. Long way to go, isn't it?

I have very strong hypotheses about this but I wait to be able to check them globally before making a decision about this subject.

---

Ok, so, first, let's summarise this into actionable things:

1. I like Tommy's suggestion, so if you agree with my comment please ignore my suggestion and apply his;
2. Given inference behaviour is likely to change in the future, I'd say it is simpler to emphasise manually setting the primary key right now. The user already had to deal with one error, best to give them an unambiguous instruction that will definitely solve the problem;
3. I just tested this and the link on the preview leads to a 404. Absolute URLs it is (only inside `errors.yaml`, `md` files should continue using relative links.

Do we have consensus on these three items? This is absolutely not a rhetorical question, please correct me if I got anything wrong.

---

> Do we need to suggest that in the error message? Or can we do that [here](https://docs.meilisearch.com/learn/core_concepts/documents.html#primary-field)?

@maryamsulemani97, I personally like the idea of giving users a hand and pointing them to a solution already on the errors page. If you prefer to do it on `core_concepts/documents`, though, I  think it's best to open another PR since this one is about error formats.

---

@gmourier thanks for the super detailed answer! It will be interesting to see what will come out of the analytics. I personally always find inference the type of thing that can help beginners but gets in the way as soon as you're trying to get some work done. As I mentioned, as a front-end dev I often received datasets over which I had no control; the thought of having to add a preprocessing step just so a third-party tool can try to infer primary keys would be really unappealing to me in that situation. Of course, these are my 0.02¥ and do not necessarily reflect Meili's users' needs.

---

Indeed @guimachiavelli, we could let the user choose his primaryKey if he doesn't want to use `id`, this way, he is not forced to modify the dataset he receives and which could have an identifier with a different name than `id`. There are several possible solutions that I think are sufficient to guide the user in the right way when he first uses it, and thus eliminate the somewhat twisted cases that we can encounter today.